File: lua/src/sekreto/json.lua
-- sekreto's own JSON, because Lua has none.
--
-- Lua 5.4's whole standard library is basic, coroutine, package, string,
-- utf8, table, math, io, os and debug - so a JSON parser is written here
-- rather than depended on. Two properties the vendor payloads need and a
-- naive table would not give: object keys keep their INSERTION ORDER (an
-- AWS payload's field order is signed, so it cannot be re-ordered by a
-- hash table), and a parse FAILURE is distinguishable from the literal
-- `null` (fetchjson must tell "this text is not JSON" from "this text is
-- the JSON null", and Lua's `nil` cannot say both).
--
-- A port of typescript's JSON handling, which is canonical.

local M = {}

-- The JSON null. A Lua nil cannot live in a table and cannot be returned
-- distinguishably, so null is a sentinel of its own.
M.NULL = setmetatable({}, { __tostring = function() return 'null' end })

local ARRMT = { __sekretojson = 'arr' }
local OBJMT = { __sekretojson = 'obj' }

-- The recursion bound. A response body arrives before any trust check has
-- been made, so `[[[[[...` must fail rather than take the process down.
local MAXDEPTH = 128

--- A JSON array, from a Lua sequence.
function M.arr(items)
  return setmetatable({ items = items or {} }, ARRMT)
end

--- A JSON object, from a list of {key, value} pairs, in that order.
function M.obj(pairlist)
  local keys = {}
  local vals = {}

  for _, pair in ipairs(pairlist or {}) do
    if nil == vals[pair[1]] then
      keys[#keys + 1] = pair[1]
    end
    vals[pair[1]] = pair[2]
  end

  return setmetatable({ keys = keys, vals = vals }, OBJMT)
end

function M.isarr(val)
  return 'table' == type(val) and ARRMT == getmetatable(val)
end

function M.isobj(val)
  return 'table' == type(val) and OBJMT == getmetatable(val)
end

function M.isnull(val)
  return M.NULL == val
end

--- Set one key on an object, keeping insertion order.
function M.set(object, key, value)
  if nil == object.vals[key] then
    object.keys[#object.keys + 1] = key
  end
  object.vals[key] = value
  return object
end

--- Walk nested objects, stopping at the first missing step. Returns nil.
function M.dig(val, ...)
  local at = val

  for _, key in ipairs({ ... }) do
    if not M.isobj(at) then
      return nil
    end
    at = at.vals[key]
    if nil == at then
      return nil
    end
  end

  return at
end

--- The value as a string, or nil for anything else.
function M.asstr(val)
  return 'string' == type(val) and val or nil
end

--- The value as a number, or nil for anything else.
function M.asnum(val)
  return 'number' == type(val) and val or nil
end

--- The value as an array's item list, or nil.
function M.asarr(val)
  return M.isarr(val) and val.items or nil
end

--- The value as an object, or nil.
function M.asobj(val)
  return M.isobj(val) and val or nil
end

--- Render a number the same way in every port: 5.0 prints as 5.
function M.numstr(val)
  if val ~= val or math.huge == val or -math.huge == val then
    return 'null'
  end

  if math.type(val) == 'integer' then
    return string.format('%d', val)
  end

  if val == math.floor(val) and math.abs(val) < 9007199254740992.0 then
    return string.format('%d', math.tointeger(val) or val)
  end

  return (string.format('%.17g', val))
end

--- Printable text, or nil. A JSON null yields nil, so a null field is a
--- miss rather than the string "null".
function M.text(val)
  if nil == val or M.NULL == val then
    return nil
  end
  if 'string' == type(val) then
    return val
  end
  if 'number' == type(val) then
    return M.numstr(val)
  end
  if 'boolean' == type(val) then
    return val and 'true' or 'false'
  end
  return nil
end

--- Render a string as a JSON string literal.
function M.quote(text)
  local out = { '"' }

  for index = 1, #text do
    local ch = text:sub(index, index)
    local byte = ch:byte()

    if '"' == ch then
      out[#out + 1] = '\\"'
    elseif '\\' == ch then
      out[#out + 1] = '\\\\'
    elseif '\n' == ch then
      out[#out + 1] = '\\n'
    elseif '\r' == ch then
      out[#out + 1] = '\\r'
    elseif '\t' == ch then
      out[#out + 1] = '\\t'
    elseif 0x20 > byte then
      out[#out + 1] = string.format('\\u%04x', byte)
    else
      out[#out + 1] = ch
    end
  end

  out[#out + 1] = '"'
  return table.concat(out)
end

--- Compact JSON text: no spaces, no newlines, keys in insertion order.
function M.stringify(val)
  if nil == val or M.NULL == val then
    return 'null'
  end

  local kind = type(val)

  if 'boolean' == kind then
    return val and 'true' or 'false'
  end

  if 'number' == kind then
    return M.numstr(val)
  end

  if 'string' == kind then
    return M.quote(val)
  end

  if M.isarr(val) then
    local parts = {}
    for index, item in ipairs(val.items) do
      parts[index] = M.stringify(item)
    end
    return '[' .. table.concat(parts, ',') .. ']'
  end

  if M.isobj(val) then
    local parts = {}
    for index, key in ipairs(val.keys) do
      parts[index] = M.quote(key) .. ':' .. M.stringify(val.vals[key])
    end
    return '{' .. table.concat(parts, ',') .. '}'
  end

  return 'null'
end

-- ---- the parser ------------------------------------------------------
--
-- Every failure below raises inside a pcall that `parse` owns, so no error
-- ever escapes it: a malformed body is answered with nil, and the caller
-- decides whether that is a miss or a failure.

local parsevalue

local WS = { [' '] = true, ['\t'] = true, ['\n'] = true, ['\r'] = true }

local function skipws(text, pos)
  local at = pos
  while at <= #text and WS[text:sub(at, at)] do
    at = at + 1
  end
  return at
end

local function bad(why)
  error({ json = true, why = why }, 0)
end

local function parseword(text, pos, word, value)
  if word ~= text:sub(pos, pos + #word - 1) then
    bad('literal')
  end
  return value, pos + #word
end

local HEX = '0123456789abcdefABCDEF'

local function parsestring(text, pos)
  if '"' ~= text:sub(pos, pos) then
    bad('string')
  end

  local at = pos + 1
  local out = {}

  while at <= #text do
    local ch = text:sub(at, at)

    if '"' == ch then
      return table.concat(out), at + 1
    end

    if '\\' ~= ch then
      out[#out + 1] = ch
      at = at + 1
    else
      local escape = text:sub(at + 1, at + 1)
      at = at + 2

      if '"' == escape then
        out[#out + 1] = '"'
      elseif '\\' == escape then
        out[#out + 1] = '\\'
      elseif '/' == escape then
        out[#out + 1] = '/'
      elseif 'b' == escape then
        out[#out + 1] = '\b'
      elseif 'f' == escape then
        out[#out + 1] = '\f'
      elseif 'n' == escape then
        out[#out + 1] = '\n'
      elseif 'r' == escape then
        out[#out + 1] = '\r'
      elseif 't' == escape then
        out[#out + 1] = '\t'
      elseif 'u' == escape then
        local digits = text:sub(at, at + 3)
        if 4 ~= #digits then
          bad('escape')
        end
        for index = 1, 4 do
          if nil == HEX:find(digits:sub(index, index), 1, true) then
            bad('escape')
          end
        end
        -- One UTF-16 code unit, appended as UTF-8. No surrogate-pair
        -- recombination, in this port or any other.
        out[#out + 1] = utf8.char(tonumber(digits, 16))
        at = at + 4
      else
        bad('escape')
      end
    end
  end

  bad('string')
end

local function parseobj(text, pos, depth)
  local at = skipws(text, pos + 1)
  local object = M.obj({})

  if '}' == text:sub(at, at) then
    return object, at + 1
  end

  while true do
    at = skipws(text, at)

    local key
    key, at = parsestring(text, at)

    at = skipws(text, at)
    if ':' ~= text:sub(at, at) then
      bad('object')
    end

    local value
    value, at = parsevalue(text, at + 1, depth + 1)

    M.set(object, key, value)

    at = skipws(text, at)
    local ch = text:sub(at, at)

    if '}' == ch then
      return object, at + 1
    end

    if ',' ~= ch then
      bad('object')
    end

    at = at + 1
  end
end

local function parsearr(text, pos, depth)
  local at = skipws(text, pos + 1)
  local items = {}

  if ']' == text:sub(at, at) then
    return M.arr(items), at + 1
  end

  while true do
    local value
    value, at = parsevalue(text, at, depth + 1)
    items[#items + 1] = value

    at = skipws(text, at)
    local ch = text:sub(at, at)

    if ']' == ch then
      return M.arr(items), at + 1
    end

    if ',' ~= ch then
      bad('array')
    end

    at = at + 1
  end
end

local NUMCHARS = '0123456789+-.eE'

local function parsenumber(text, pos)
  local at = pos

  while at <= #text and nil ~= NUMCHARS:find(text:sub(at, at), 1, true) do
    at = at + 1
  end

  local value = tonumber(text:sub(pos, at - 1))

  -- 1e999 parses to infinity, which JSON has no literal for and which
  -- would later blow up a token-expiry computation.
  if nil == value or value ~= value or math.huge == value or -math.huge == value then
    bad('number')
  end

  return value, at
end

parsevalue = function(text, pos, depth)
  if MAXDEPTH < depth then
    bad('depth')
  end

  local at = skipws(text, pos)
  local ch = text:sub(at, at)

  if '' == ch then
    bad('eof')
  end

  if '{' == ch then
    return parseobj(text, at, depth)
  end

  if '[' == ch then
    return parsearr(text, at, depth)
  end

  if '"' == ch then
    return parsestring(text, at)
  end

  if 't' == ch then
    return parseword(text, at, 'true', true)
  end

  if 'f' == ch then
    return parseword(text, at, 'false', false)
  end

  if 'n' == ch then
    return parseword(text, at, 'null', M.NULL)
  end

  return parsenumber(text, at)
end

--- Parse JSON text. Returns the value, or nil when the text is not JSON.
--- The literal `null` parses to M.NULL, which is NOT nil - that is the
--- whole reason the sentinel exists.
function M.parse(text)
  if 'string' ~= type(text) or '' == text then
    return nil
  end

  local ok, value, at = pcall(parsevalue, text, 1, 1)
  if not ok then
    return nil
  end

  if #text >= skipws(text, at) then
    return nil
  end

  return value
end

return M

File: haskell/plugins/Json.hs
-- | Minimal JSON support for sekreto.
--
-- sekreto adds no third-party dependencies, so it carries just enough JSON
-- to read a vault's answer and write the CLI's own line of output. It is
-- deliberately not a general-purpose library.
--
-- 'parse' answers @Maybe Json@, where 'Nothing' means "this text is not
-- JSON" and @Just JNull@ means "this text is the JSON literal null" - a
-- distinction the callers of fetchjson need, since only the first is a
-- malformed response.
--
-- Objects keep their fields in the order they arrived: a payload's field
-- order is signed, and `Data.Map` orders by key.
--
-- A port of typescript/src/Json.ts, which is canonical.

module Json
  ( Json (..),
    asarr,
    asnum,
    asobj,
    asstr,
    dig,
    numstr,
    parse,
    quote,
    stringify,
    text,
  )
where

import Data.Char (chr, isDigit, isHexDigit, isSpace, ord)
import Data.List (foldl')
import Numeric (readHex, showHex)

-- | A JSON value. Numbers are 'Double' only, as in the canonical port;
-- objects are an association list, so the order fields arrived in is the
-- order they leave in.
data Json
  = JNull
  | JBool Bool
  | JNum Double
  | JStr String
  | JArr [Json]
  | JObj [(String, Json)]
  deriving (Eq, Show)

asstr :: Maybe Json -> Maybe String
asstr (Just (JStr value)) = Just value
asstr _ = Nothing

asnum :: Maybe Json -> Maybe Double
asnum (Just (JNum value)) = Just value
asnum _ = Nothing

asarr :: Maybe Json -> Maybe [Json]
asarr (Just (JArr value)) = Just value
asarr _ = Nothing

asobj :: Maybe Json -> Maybe [(String, Json)]
asobj (Just (JObj value)) = Just value
asobj _ = Nothing

-- | Walk nested objects, stopping the moment a step is not there. The
-- argument is optional because a response body is optional: a store may
-- not have answered with JSON at all.
dig :: Maybe Json -> [String] -> Maybe Json
dig = foldl' step
  where
    step (Just (JObj entries)) key = lookup key entries
    step _ _ = Nothing

-- | This value as the text a caller would print, or nothing at all when
-- there is no value. A JSON null is "no value": every provider here treats
-- it as a miss rather than as the string @null@.
text :: Maybe Json -> Maybe String
text Nothing = Nothing
text (Just JNull) = Nothing
text (Just (JStr value)) = Just value
text (Just (JNum value)) = Just (numstr value)
text (Just (JBool value)) = Just (if value then "true" else "false")
text (Just other) = Just (stringify other)

-- | Render a number the way every other port does: a whole number has no
-- fractional tail, so a JSON @1@ read back and printed stays @1@.
numstr :: Double -> String
numstr value
  | isNaN value || isInfinite value = "null"
  | value == fromIntegral rounded && 9007199254740992.0 > abs value = show rounded
  | otherwise = show value
  where
    rounded = truncate value :: Integer

-- | Render a string as a JSON string literal, quotes included. Public,
-- because the CLI assembles its output line field by field.
quote :: String -> String
quote body = '"' : concatMap escape body ++ "\""
  where
    escape '"' = "\\\""
    escape '\\' = "\\\\"
    escape '\n' = "\\n"
    escape '\r' = "\\r"
    escape '\t' = "\\t"
    escape head
      | 0x20 > ord head = "\\u" ++ pad (showHex (ord head) "")
      | otherwise = [head]

    pad digits = replicate (4 - length digits) '0' ++ digits

-- | Render a value as compact JSON: no spaces, no newlines.
stringify :: Json -> String
stringify JNull = "null"
stringify (JBool value) = if value then "true" else "false"
stringify (JNum value) = numstr value
stringify (JStr value) = quote value
stringify (JArr entries) = "[" ++ joined (map stringify entries) ++ "]"
stringify (JObj entries) = "{" ++ joined (map field entries) ++ "}"
  where
    field (key, value) = quote key ++ ":" ++ stringify value

joined :: [String] -> String
joined [] = ""
joined [one] = one
joined (head : rest) = head ++ "," ++ joined rest

-- | Parse JSON text. 'Nothing' for anything unreadable - which the caller
-- must tell apart from a literal @null@ body, since only the first means
-- the store could not answer coherently.
--
-- No error escapes: every failure is 'Nothing'. Nesting is capped, because
-- a response body arrives before any trust check and @[[[[...@ must not
-- take the stack down.
parse :: String -> Maybe Json
parse body
  | null body = Nothing
  | otherwise = case value 0 (skip body) of
      Just (parsed, rest) | null (skip rest) -> Just parsed
      _ -> Nothing

maxdepth :: Int
maxdepth = 128

skip :: String -> String
skip = dropWhile isSpace

value :: Int -> String -> Maybe (Json, String)
value depth body
  | maxdepth < depth = Nothing
  | otherwise = case body of
      ('{' : rest) -> object depth (skip rest) []
      ('[' : rest) -> array depth (skip rest) []
      ('"' : rest) -> fmap (\(parsed, left) -> (JStr parsed, left)) (jstring rest)
      _ -> literal body
  where
    literal text'
      | "true" == take 4 text' = Just (JBool True, drop 4 text')
      | "false" == take 5 text' = Just (JBool False, drop 5 text')
      | "null" == take 4 text' = Just (JNull, drop 4 text')
      | otherwise = number text'

object :: Int -> String -> [(String, Json)] -> Maybe (Json, String)
object _ ('}' : rest) sofar = Just (JObj (reverse sofar), rest)
object depth body sofar = do
  (key, afterkey) <- case body of
    ('"' : rest) -> jstring rest
    _ -> Nothing

  rest <- case skip afterkey of
    (':' : left) -> Just left
    _ -> Nothing

  (entry, afterentry) <- value (depth + 1) (skip rest)

  -- A later duplicate overwrites the earlier one rather than sitting
  -- beside it, so that `lookup` cannot answer with the value the writer
  -- replaced.
  let kept = (key, entry) : filter ((key /=) . fst) sofar

  case skip afterentry of
    (',' : left) -> object depth (skip left) kept
    ('}' : left) -> Just (JObj (reverse kept), left)
    _ -> Nothing

array :: Int -> String -> [Json] -> Maybe (Json, String)
array _ (']' : rest) sofar = Just (JArr (reverse sofar), rest)
array depth body sofar = do
  (entry, afterentry) <- value (depth + 1) body

  case skip afterentry of
    (',' : left) -> array depth (skip left) (entry : sofar)
    (']' : left) -> Just (JArr (reverse (entry : sofar)), left)
    _ -> Nothing

-- | The body of a string, the opening quote already taken.
jstring :: String -> Maybe (String, String)
jstring = walk ""
  where
    walk sofar ('"' : rest) = Just (reverse sofar, rest)
    walk _ [] = Nothing
    walk sofar ('\\' : escape : rest) = case escape of
      '"' -> walk ('"' : sofar) rest
      '\\' -> walk ('\\' : sofar) rest
      '/' -> walk ('/' : sofar) rest
      'b' -> walk ('\b' : sofar) rest
      'f' -> walk ('\f' : sofar) rest
      'n' -> walk ('\n' : sofar) rest
      'r' -> walk ('\r' : sofar) rest
      't' -> walk ('\t' : sofar) rest
      -- Exactly four hex digits, appended as one UTF-16 code unit: no
      -- port recombines surrogate pairs.
      'u' ->
        let digits = take 4 rest
         in if 4 == length digits && all isHexDigit digits
              then case readHex digits of
                [(code, "")] -> walk (chr code : sofar) (drop 4 rest)
                _ -> Nothing
              else Nothing
      _ -> Nothing
    walk sofar (head : rest) = walk (head : sofar) rest

-- | A number, handed to the platform's float reader. A non-finite result -
-- @1e999@ parses to infinity - is refused: JSON has no infinity, and one
-- reaching a token-expiry computation blows it up later.
number :: String -> Maybe (Json, String)
number body
  | null span' = Nothing
  | otherwise = case reads (fixup span') :: [(Double, String)] of
      [(parsed, "")] | not (isNaN parsed) && not (isInfinite parsed) -> Just (JNum parsed, rest)
      _ -> Nothing
  where
    (span', rest) = Prelude.span numeric body
    numeric head = isDigit head || elem head "+-.eE"

    -- `reads` for Double will not take a leading '+' or '.', nor a bare
    -- trailing '.', so the span is normalised before it is handed over.
    fixup text' = case dropWhile ('+' ==) text' of
      ('-' : left) -> '-' : mantissa left
      left -> mantissa left

    mantissa text' = case text' of
      ('.' : left) -> "0." ++ left
      left -> left
